src/portgraph.rs

```diff
@@ -145,35 +144,56 @@ impl PortGraph {
         }
     }
 
+    /// Allocates a slab of ports. Returns the index of the first port, and the
+    /// number of ports in the slab.
     #[inline]
-    fn alloc_ports(&mut self, node: NodeIndex, incoming: usize, outgoing: usize) -> PortIndex {
-        let size = incoming + outgoing;
+    fn alloc_ports(
+        &mut self,
+        node: NodeIndex,
+        incoming: usize,
+        outgoing: usize,
+        extra_capacity: usize,
+    ) -> NodeMeta {
+        let requested = incoming + outgoing + extra_capacity;
         let meta_incoming = PortEntry::Port(PortMeta::new(node, Direction::Incoming));
         let meta_outgoing = PortEntry::Port(PortMeta::new(node, Direction::Outgoing));
+        let meta_empty = PortEntry::Free;
+
+        if requested == 0 {
+            return NodeMeta::default();
+        }
 
-        match self.port_free.get(size - 1).copied().flatten() {
+        match self.port_free.get(requested - 1).copied().flatten() {
             Some(port) => {
+                // TODO: Over-allocate if there are similarly-sized free slabs.
+                let capacity = requested;
+
                 let free_next = take(&mut self.port_link[port.index()]);
-                self.port_free[size - 1] = free_next;
+                self.port_free[requested - 1] = free_next;
 
-                let i = port.index();
-                self.port_meta[i + incoming..i + size].fill(meta_outgoing);
-                self.port_meta[i..i + incoming].fill(meta_incoming);
-                self.port_link[i..i + size].fill(None);
+                let node_meta =
+                    NodeMeta::new(port, incoming as u16, outgoing as u16, capacity as u16);
+                self.port_meta[node_meta.incoming_ports()].fill(meta_incoming);
+                self.port_meta[node_meta.outgoing_ports()].fill(meta_outgoing);
+                self.port_meta[node_meta.unused_ports()].fill(meta_empty);
+                self.port_link[port.index()..port.index() + capacity].fill(None);
 
-                port
+                node_meta
             }
             None => {
                 debug_assert_eq!(self.port_meta.len(), self.port_link.len());
                 let old_len = self.port_meta.len();
                 let port = PortIndex::new(old_len);
+                let capacity = requested;
 
-                self.port_meta.reserve(size);
+                self.port_meta.reserve(requested);
                 self.port_meta.resize(old_len + incoming, meta_incoming);
-                self.port_meta.resize(old_len + size, meta_outgoing);
-                self.port_link.resize(old_len + size, None);
+                self.port_meta
+                    .resize(old_len + incoming + outgoing, meta_outgoing);
+                self.port_meta.resize(old_len + capacity, meta_empty);
+                self.port_link.resize(old_len + capacity, None);
```

Comment:
why do you need the second line?

Reply:
`port_meta` vs `port_link`. The second has the port connections (all disconnected at allocation time)

Reply:
ah sorry